category_theory/natural_transformation.lean

```diff
@@ -0,0 +1,117 @@
+/-
+Copyright (c) 2017 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Tim Baumann, Stephen Morgan, Scott Morrison
+
+Defines natural transformations between functors.
+
+Introduces notations
+  `F ⟹ G` for the type of natural transformations between functors `F` and `G`,
+  `τ X` (a coercion) for the components of natural transformations,
+  `σ ⊟ τ` for vertical compositions, and
+  `σ ◫ τ` for horizontal compositions.
+-/
+
+import .functor
+
+namespace category_theory
+
+universes u₁ v₁ u₂ v₂ u₃ v₃
+
+variables {C : Type u₁} [𝒞 : category.{u₁ v₁} C] {D : Type u₂} [𝒟 : category.{u₂ v₂} D]
+include 𝒞 𝒟
+
+/--
+`nat_trans F G` represents a natural transformation between functors `F` and `G`.
+
+The field `app` provides the components of the natural transformation, and there is a
+coercion available so you can write `α X` for the component of a transformation `α` at an object `X`.
+
+Naturality is expressed by `α.naturality_lemma`.
+-/
+structure nat_trans (F G : C ↝ D) : Type (max u₁ v₂) :=
+(app : Π X : C, (F X) ⟶ (G X))
+(naturality : ∀ {X Y : C} (f : X ⟶ Y), (F.map f) ≫ (app Y) = (app X) ≫ (G.map f) . obviously)
+
+restate_axiom nat_trans.naturality
+attribute [ematch] nat_trans.naturality_lemma
+
+infixr ` ⟹ `:50  := nat_trans             -- type as \==> or ⟹
+
+namespace nat_trans
+
+instance {F G : C ↝ D} : has_coe_to_fun (F ⟹ G) :=
+{ F   := λ α, Π X : C, (F X) ⟶ (G X),
+  coe := λ α, α.app }
+
+@[simp] lemma coe_def {F G : C ↝ D} (α : F ⟹ G) (X : C) : α X = α.app X := rfl
+
+/-- `nat_trans.id F` is the identity natural transformation on a functor `F`. -/
+protected definition id (F : C ↝ D) : F ⟹ F := 
+{ app        := λ X, 𝟙 (F X),
+  naturality := begin /- `obviously'` says: -/ intros, dsimp, simp end }
+
+@[simp] lemma id.app (F : C ↝ D) (X : C) : (nat_trans.id F) X = 𝟙 (F X) := rfl
```

Comment:
we usually use `_` instead of `.` for theorems referring to a definition unless it makes sense with projection notation

Reply:
Fixed!